### PR TITLE
Async issues, code cleanup and some error handling

### PR DESCRIPTION
**Pull request by sspanak**
I haven't changed the package versions in `package.json` as I can't publish anything to NPM. These need to be manually adjusted after publishing.

### CODE REVIEW

File: camera-processor/src/CameraProcessor.ts [hunk 1]
```diff
@@ -14,14 +14,11 @@ class CameraProcessor<TAnalyzerData = any> {
   public passthrough: boolean = false;
   public isRunning: boolean = false;
 
-  constructor() {
-    this.processFrame = this.processFrame.bind(this); // For callback purposes
```
Comment (spasimir21) [inline camera-processor/src/CameraProcessor.ts:18]: This way we don't have to create a new function for every frame.

File: camera-processor/src/CameraProcessor.ts [hunk 2]
```diff
@@ -14,14 +14,11 @@ class CameraProcessor<TAnalyzerData = any> {
...
-  }
 
-  start(): void {
+  async start(): Promise<void> {
+    await this.cameraVideo.play();
```
Comment (spasimir21) [inline camera-processor/src/CameraProcessor.ts:19]: Browsers before 2019 won't return a value for play, so we need to handle this with readyState.

File: camera-processor/src/CameraProcessor.ts [hunk 3]
```diff
@@ -32,7 +29,6 @@ class CameraProcessor<TAnalyzerData = any> {
   setCameraStream(stream: MediaStream): void {
     this.cameraStream = stream;
     this.cameraVideo.srcObject = this.cameraStream;
-    this.cameraVideo.play();
```
Comment (spasimir21) [inline camera-processor/src/CameraProcessor.ts:35]: This is needed for the video to refresh.
Could probably be handled better though.
Should check isRunning.

File: camera-processor/src/CameraProcessor.ts [hunk 4]
```diff
@@ -14,14 +14,11 @@ class CameraProcessor<TAnalyzerData = any> {
   public passthrough: boolean = false;
   public isRunning: boolean = false;
 
-  constructor() {
```
Comment (spasimir21) [inline camera-processor/src/CameraProcessor.ts:17]: We also need to mute the video for autoplay reasons.

File: camera-processor/src/CameraProcessor.ts [hunk 5]
```diff
@@ -59,8 +55,12 @@ class CameraProcessor<TAnalyzerData = any> {
     this.performance.frameTime.analyze = time_analyze - time_start;
     this.performance.frameTime.render = time_render - time_analyze;
     this.performance.frameTime.total = time_render - time_start;
-    this.performance.fps = Math.min(1000 / this.performance.frameTime.total, 1000);
-    if (this.isRunning) requestAnimationFrame(this.processFrame);
+
+    // Time precision of Date.now() is not up to a millisecond for security reasons, so we may end up having zero frame time.
+    // More info: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/now#reduced_time_precision
+    this.performance.fps = this.performance.frameTime.total > 0 ? this.performance.frameTime.total * 0.001 : 1000;
```
Comment (spasimir21) [inline camera-processor/src/CameraProcessor.ts:61]: 1000 / x != x * 0.001
Should be: this.performance.frameTime.total > 0 ? 1000 / this.performance.frameTime.total : 1000;
But I would prefer to keep the original solution.

File: @camera-processor/virtual-background/src/SegmentationAnalyzer.ts [hunk 6]
```diff
@@ -5,49 +5,43 @@ import CameraProcessor, { FrameAnalyzer } from 'camera-processor';
 import { SEGMENTATION_BACKEND } from './shared';
 
 class SegmentationAnalyzer extends FrameAnalyzer {
-  private segmentationResult: { width: number; height: number; data: any };
   private readonly backends: { [key: string]: SegmentationBackend } = {};
-  public backendId: SEGMENTATION_BACKEND;
-  public backend: SegmentationBackend;
-  public segmentationSettings: any;
-  public modelSettings: any;
```
Comment (spasimir21) [inline @camera-processor/virtual-background/src/SegmentationAnalyzer.ts:13]: Segmentation and model settings should be readable from the top class.
Should be made getters.

File: @camera-processor/virtual-background/src/VirtualBackgroundRenderer.ts [hunk 7]
```diff
@@ -7,14 +7,17 @@ import RenderPipeline from './render_pipeline/RenderPipeline';
 class VirtualBackgroundRenderer extends FrameRenderer {
   public readonly background: VirtualBackground = { type: VIRTUAL_BACKGROUND_TYPE.None, data: null };
   private readonly pipelines: { [key: string]: RenderPipeline } = {};
-  public pipelineId: RENDER_PIPELINE;
-  public pipeline: RenderPipeline;
-  public renderSettings: any;
```
Comment (spasimir21) [inline @camera-processor/virtual-background/src/VirtualBackgroundRenderer.ts:12]: Same as the SegmentationAnalyzer.

File: camera-processor/src/CameraProcessor.ts (hunk 5 above)
Comment (sspanak) [reply camera-processor/src/CameraProcessor.ts:61]: Yes, my bad, I have turned the expression upside down. Anyway, using the ternary operator should yield faster performance
```suggestion
    this.performance.fps = this.performance.frameTime.total > 0 ? 1000 / this.performance.frameTime.total : 1000;
```

**Comment by spasimir21** [comment]
PR was committed with small changes.

**Pull request closed without merging**